frontends/ast/ast.h

```diff
@@ -321,6 +321,9 @@ namespace AST
 		static AstNode *mkconst_str(const std::vector<RTLIL::State> &v);
 		static AstNode *mkconst_str(const std::string &str);
 
+		// helper function to create an AST node for a temporary register
+		AstNode *mktemp_logic(const std::string &name, AstNode *mod, bool nosync, int range_left, int range_right, bool is_signed);
```

Comment:
It can't be static, as long as it's encoding the filename and location in `wire->str`.

Is that a problem? If so, I guess I could make a separate (non-static) function to generate the name, and pass the result of that into `mktemp_logic`, but I'm not sure I see the benefit.

Reply:
Ah sorry, missed that. No, it's not a problem per se

Reply:
FYI I've also thought about enhancing the function later, adding a parameter to decide whether the name should be local (no name decoration) or global (name decorated as it is now).

I believe the function could then be used to replace most of the current code to instantiate temporaries, for which I could make a PR.

Reply:
@povik As promised did some more tests with shift_issue3.zip (including reg_demux_noshift in the Makefile). I applied #3877 to reproduce increased chip area when an unconditional temporary was introduced for the rhs.

Without any changes (no temporary in this case):
```   Chip area for module '\reg_demux_noshift': 18689.907600```
Forcing use of the temporary, which is currently assigned by AST_ASSIGN_EQ in the current block, instantiating a `(* nosync  *)` register:
```   Chip area for module '\reg_demux_noshift': 18815.101200```
Forcing use of the temporary using AST_ASSIGN at the module level instead, instantiating a wire, since it could possibly be argued that this should use less resources than a register in this case (it did the contrary):
```   Chip area for module '\reg_demux_noshift': 18896.749200```

As it stands for this test case, I'm hesitant to make any further changes.

Perhaps someone wants to investigate why introducing a temporary makes any difference at all (e.g. should it have been completely optimized away in `simplify.cc` / `genrtlil.cc` in this case, since it's assigned from a simple 117 bit wire?), however this is outside the scope for this PR.

Reply:
Thanks! I experimented with this, and as far as I can tell the difference in area is due to the chaotic nature of the synthesis flow, not because of any systematic effect of introducing the temporary. When that's the case I think we want to make the temporary non-conditional, just so we spare a bit of control flow in the congested `simplify.cc` code.

Reply:
OK, done.